Reimplements the corDapp template in Java.

joeldudleyr3 opened the pull request.

shamsasari commented on java/src/main/java/com/example/contract/PurchaseOrderContract.java:
```diff
@@ -0,0 +1,147 @@
+                        command.getSigners(),
+                        command.getSigningParties(),
+                        (Commands) command.getValue()))
+                .collect(Collectors.toList());
```
static import

joeldudleyr3 replied:
Done

shamsasari commented on java/src/main/java/com/example/contract/PurchaseOrderContract.java:
```diff
@@ -0,0 +1,147 @@
+                List<? extends ContractState> inputs,
+                List<? extends ContractState> outputs,
+                List<? extends AuthenticatedObject<? extends Commands>> commands,
+                Unit groupingKey) {
```
Move `{` to the next line

joeldudleyr3 replied:
Done

shamsasari commented on java/src/main/java/com/example/contract/PurchaseOrderContract.java:
```diff
@@ -0,0 +1,147 @@
+                List<? extends ContractState> outputs,
+                List<? extends AuthenticatedObject<? extends Commands>> commands,
+                Unit groupingKey) {
+
```
Remove blank line

joeldudleyr3 replied:
Done

shamsasari commented on java/src/main/java/com/example/contract/PurchaseOrderContract.java:
```diff
@@ -0,0 +1,147 @@
+
+        // If you add additional clauses, make sure to reference them within the 'AnyComposition()' clause.
+        class Group extends GroupClauseVerifier<PurchaseOrderState, Commands, UniqueIdentifier> {
+
```
Remove blank line

joeldudleyr3 replied:
Done

shamsasari commented on java/src/main/java/com/example/contract/PurchaseOrderContract.java:
```diff
@@ -0,0 +1,147 @@
+                Unit groupingKey) {
+
+                if (tx.getTimestamp() == null) {
+                    throw new IllegalArgumentException("must be timestamped");
```
You could use `Objects.requireNotNull` or `Preconditions.checkArgument`

shamsasari commented on java/src/main/java/com/example/contract/PurchaseOrderContract.java:
```diff
@@ -0,0 +1,147 @@
+                    require.by("The delivery date must be in the future.",
+                            out.getPo().getDeliveryDate().toInstant().isAfter(time));
+
+                    return Unit.INSTANCE;
```
Return `null` would be cleaner

joeldudleyr3 replied:
Done

shamsasari commented on java/src/main/java/com/example/contract/PurchaseOrderState.java:
```diff
@@ -0,0 +1,81 @@
+        return getParties()
+                .stream()
+                .map(Party::getOwningKey)
+                .collect(Collectors.toList());
```
static import

joeldudleyr3 replied:
Done

shamsasari commented on java/src/main/java/com/example/contract/PurchaseOrderState.java:
```diff
@@ -0,0 +1,81 @@
+ * interfaces for different requirements. In this case, we are implementing a [DealState] which defines a few helper
+ * properties and methods for managing states pertaining to deals. */
+public class PurchaseOrderState implements DealState {
+    private PurchaseOrder po;
```
finals

shamsasari commented on java/src/main/java/com/example/contract/PurchaseOrderState.java:
```diff
@@ -0,0 +1,81 @@
+                .collect(Collectors.toList());
+        return !(ourKeys
+                .stream()
+                .filter(partyKeys::contains)
```
You can use `anyMatch` here

shamsasari commented on java/src/main/java/com/example/contract/PurchaseOrderState.java:
```diff
@@ -0,0 +1,81 @@
+        List<PublicKey> partyKeys = getParties()
+                .stream()
+                .flatMap(party -> party.getOwningKey().getKeys().stream())
+                .collect(Collectors.toList());
```
You're collecting into a List and then re-opening a new stream on that - just continue the Stream chain.

shamsasari commented on java/src/main/java/com/example/contract/PurchaseOrderState.java:
```diff
@@ -0,0 +1,81 @@
+     * https://docs.corda.net/api/net.corda.core.transactions/-transaction-builder/index.html */
+    @Override public TransactionBuilder generateAgreement(Party notary) {
+        return new TransactionType.General.Builder(notary)
+                .withItems(this, new Command(new PurchaseOrderContract.Commands.Place(), getParticipants()));
```
Perhaps static import `Commands` so that you get `Commands.Place()`. I imagine it should be obvious we're dealing with our own commands.

joeldudleyr3 replied:
Done

shamsasari commented on java/src/main/java/com/example/flow/ExampleFlow.java:
```diff
@@ -0,0 +1,259 @@
+        private Party otherParty;
+        private ProgressTracker progressTracker;
+
+        private ProgressTracker.Step CONSTRUCTING_OFFER = new ProgressTracker.Step(
```
Make all of these steps `static final`

joeldudleyr3 replied:
Done

shamsasari commented on java/src/main/java/com/example/flow/ExampleFlow.java:
```diff
@@ -0,0 +1,259 @@
+        private ProgressTracker.Step SENDING_FINAL_TRANSACTION = new ProgressTracker.Step(
+                "Sending fully signed transaction to seller.");
+
+        public Initiator(DealState po, Party otherParty) {
```
This is a `PurchaseOrder`? Should you not have it of that type?

joeldudleyr3 replied:
Good point

shamsasari commented on java/src/main/java/com/example/flow/ExampleFlow.java:
```diff
@@ -0,0 +1,259 @@
+                // Flow jumps to Acceptor.
+                // -----------------------
+                // Receive the partially signed transaction off the wire from the other party.
+                SignedTransaction ptx = this.sendAndReceive(otherParty, offerMessage, SignedTransaction.class)
```
Can you add a TODO to have the `Class` param at the front.

shamsasari replied:
`this` isn't needed

joeldudleyr3 replied:
Will do

joeldudleyr3 replied:
Done

shamsasari commented on java/src/main/java/com/example/flow/ExampleFlow.java:
```diff
@@ -0,0 +1,259 @@
+        private Party otherParty;
+        private ProgressTracker progressTracker;
+
+        private ProgressTracker.Step WAIT_FOR_AND_RECEIVE_PROPOSAL = new ProgressTracker.Step("Receiving proposed purchase order from buyer.");
```
static finals

shamsasari commented on java/src/main/java/com/example/flow/ExampleFlow.java:
```diff
@@ -0,0 +1,259 @@
+                Instant currentTime = getServiceHub().getClock().instant();
+                // As we are running in a distributed system, we allocate a 30 second time window for the transaction to
+                // be timestamped by the Notary service.
+                utx.setTime(currentTime, Utils.getSeconds(30));
```
This `Utils` method only makes sense with Kotlin syntax. In Java land you're better off with the original - `Duration.ofDays`.

Also can I ask you to add a TODO to `Utils` to separate out the functions useful just for Kotlin b/c most of the stuff there isn't useful in Java.

joeldudleyr3 replied:
Done and will do

shamsasari commented on java/src/main/java/com/example/flow/ExampleFlow.java:
```diff
@@ -0,0 +1,259 @@
+                // Record the transaction.
+                progressTracker.setCurrentStep(RECORDING);
+                getServiceHub().recordTransactions(Collections.singletonList(ntx));
+                return new ExampleFlowResult.Success(String.format("Transaction id %s committed to ledger.", ntx.getId().toString()));
```
I don't think you need to call toString() - `ntx.getId()` should be fine.

joeldudleyr3 replied:
Done

shamsasari commented on java/src/main/java/com/example/flow/ExampleFlow.java:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    public static class ExampleFlowResult {
+        static public class Success extends com.example.flow.ExampleFlow.ExampleFlowResult {
```
`public static`

joeldudleyr3 replied:
Done

shamsasari commented on java/src/main/java/com/example/flow/ExampleFlow.java:
```diff
@@ -0,0 +1,259 @@
+            public String toString() { return String.format("Success(%s)", message); }
+        }
+
+        static public class Failure extends com.example.flow.ExampleFlow.ExampleFlowResult {
```
likewise

joeldudleyr3 replied:
Done

shamsasari commented on java/src/main/java/com/example/model/PurchaseOrder.java:
```diff
@@ -0,0 +1,61 @@
+ * separate file. */
+
+/** A simple class representing a purchase order. */
+public class PurchaseOrder {
```
Where art thou data classes?!

shamsasari commented on java/src/main/java/com/example/model/PurchaseOrder.java:
```diff
@@ -0,0 +1,61 @@
+                orderNumber,
+                deliveryDate.toString(),
+                deliveryAddress.toString(),
+                java.util.Arrays.toString(items.toArray()));
```
import `Arrays`

shamsasari commented on java/src/main/java/com/example/plugin/ExamplePlugin.java:
```diff
@@ -0,0 +1,64 @@
+    /** A list of directories in the resources directory that will be served by Jetty under /web. */
+    private Map<String, String> staticServeDirs = Collections.singletonMap(
+            // This will serve the exampleWeb directory in resources to /web/example
+            "example", this.getClass().getClassLoader().getResource("exampleWeb").toExternalForm()
```
`this` isn't really necessary

shamsasari commented on java/src/main/java/com/example/plugin/ExamplePlugin.java:
```diff
@@ -0,0 +1,64 @@
+    );
+
+    public ExamplePlugin() {
+        Set<String> classNames = new HashSet<>(Arrays.asList(
```
Is this the param types of the flow?

joeldudleyr3 replied:
It's a list of whitelisted flows and arguments.

I've removed the public constructor now - seemed easier.

shamsasari commented on java/src/main/python/ExampleClientRPC.py:
```diff
@@ -0,0 +1,30 @@
+    for txn in txs:
+        print(txn.tx.outputs[0].data.po)
+
+
```
Remove trailing blank lines

joeldudleyr3 replied:
Done.

shamsasari commented on java/src/test/java/com/example/ExampleTest.java:
```diff
@@ -0,0 +1,10 @@
+
+import org.junit.Test;
+
+public class ExampleTest {
```
This is not doing much!

joeldudleyr3 replied:
Yes - but this is the case in the Kotlin version too. There's a sprint issue to add unit tests, which will fix this.

joeldudleyr3 commented on README.md:
```diff
@@ -19,16 +19,20 @@ orders to other nodes. You can also enumerate all the purchase orders
 which have been agreed with other nodes. The nodes also provide a simple
 web interface which can be used to interact with the CorDapp.
 
+The source code for this CorDapp is provided in both Kotlin (under /kotlin) 
```
Done

joeldudleyr3 commented on java/src/main/java/com/example/Main.java:
```diff
@@ -0,0 +1,51 @@
+    public static void main(String[] args) {
+        // No permissions required as we are not invoking flows.
+        User user = new User("user1", "test", Collections.emptySet());
+        Driver.driver(
```
Done

joeldudleyr3 commented on java/src/main/java/com/example/Main.java:
```diff
@@ -0,0 +1,51 @@
+                new PortAllocation.Incremental(5005),
+                false,
+                true,
+                dsl -> {dsl.startNode("Controller",
```
Done

joeldudleyr3 commented on java/src/main/java/com/example/Main.java:
```diff
@@ -0,0 +1,51 @@
+                        Collections.singleton(new ServiceInfo(ValidatingNotaryService.Companion.getType(), null)),
+                        Collections.emptyList(),
+                        Collections.emptyMap());
+                    dsl.startNode("NodeA", Collections.emptySet(), Collections.singletonList(user), Collections.emptyMap());
```
Done

joeldudleyr3 commented on java/src/main/java/com/example/Main.java:
```diff
@@ -0,0 +1,51 @@
+                    dsl.startNode("NodeB", Collections.emptySet(), Collections.singletonList(user), Collections.emptyMap());
+                    dsl.startNode("NodeC", Collections.emptySet(), Collections.singletonList(user), Collections.emptyMap());
+                    dsl.waitForAllNodesToFinish();
+                    return Unit.INSTANCE; }
```
Done

joeldudleyr3 replied:
Done

joeldudleyr3 commented on java/src/main/java/com/example/api/ExampleApi.java:
```diff
@@ -0,0 +1,103 @@
+        return Collections.singletonMap("me", me);
+    }
+
+    /** Returns all parties registered with the [NetworkMapService]. The names can be used to look up identities by
```
Done

joeldudleyr3 commented on java/src/main/java/com/example/api/ExampleApi.java:
```diff
@@ -0,0 +1,103 @@
+                services.getNetworkMapCache().getPartyNodes()
+                        .stream()
+                        .map(node -> node.getLegalIdentity().getName())
+                        .filter(name -> !name.equals(me) && !name.equals("Controller"))
```
Done

joeldudleyr3 commented on java/src/main/java/com/example/api/ExampleApi.java:
```diff
@@ -0,0 +1,103 @@
+                        .stream()
+                        .map(node -> node.getLegalIdentity().getName())
+                        .filter(name -> !name.equals(me) && !name.equals("Controller"))
+                        .collect(Collectors.toList()));
```
Done

joeldudleyr3 commented on java/src/main/java/com/example/api/ExampleApi.java:
```diff
@@ -0,0 +1,103 @@
+    @Path("purchase-orders")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Map<UniqueIdentifier, StateAndRef<PurchaseOrderState>> getPurchaseOrders() {
+        return services.getVaultService().linearHeadsOfType_(PurchaseOrderState.class);
```
That's the name in the original Kotlin file, seemingly to prevent a clash with the inlined linearHeadsOfType().

joeldudleyr3 commented on java/src/main/java/com/example/api/ExampleApi.java:
```diff
@@ -0,0 +1,103 @@
+     * The flow is invoked asynchronously. It returns a future when the flow's call() method returns. */
+    @PUT
+    @Path("{party}/create-purchase-order")
+    public Response createPurchaseOrder(PurchaseOrder po, @PathParam("party") String partyName) throws InterruptedException, ExecutionException {
```
Done

joeldudleyr3 commented on java/src/main/java/com/example/api/ExampleApi.java:
```diff
@@ -0,0 +1,103 @@
+                    state,
+                    otherParty).getResultFuture().get();
+
+            if (result instanceof ExampleFlow.ExampleFlowResult.Success) {
```
Done

joeldudleyr3 commented on java/src/main/java/com/example/api/ExampleApi.java:
```diff
@@ -0,0 +1,103 @@
+                        .build();
+            }
+        } else {
+            return Response.status(Response.Status.BAD_REQUEST).build();
```
Done

joeldudleyr3 commented on java/src/main/java/com/example/client/ExampleClientRPC.java:
```diff
@@ -0,0 +1,54 @@
+                                })
+        );
+
+        new CompletableFuture<Unit>().get();
```
Done

joeldudleyr3 commented:
Hi both,

Shams, I've addressed your comments. In places, Shams's comments applied to the Kotlin version as well, so I've made changes there too.

I also blew away my local maven, reinstalled the latest version of corda-master, and changed the template where required to match the latest changes.

Roger, can you check the final version?

shamsasari commented on java/src/main/java/com/example/api/ExampleApi.java:
```diff
@@ -4,100 +4,110 @@
-    private ServiceHub services;
-    private String me;
+    private final CordaRPCOps services;
+    private final String me;
```
myLegalName would be better

joeldudleyr3 replied:
Done

shamsasari commented on java/src/main/java/com/example/Main.java:
```diff
@@ -26,26 +25,29 @@
         // No permissions required as we are not invoking flows.
-        User user = new User("user1", "test", Collections.emptySet());
-        Driver.driver(
+        final User user = new User("user1", "test", emptySet());
```
FYI: final local variables is not universal in Java purely because it clutters the code, but you can keep it if your prefer it.

shamsasari commented on java/src/main/java/com/example/api/ExampleApi.java:
```diff
@@ -4,100 +4,110 @@
-                    ExampleFlow.Initiator.class,
-                    state,
-                    otherParty).getResultFuture().get();
+        // The line below blocks and waits for the future to resolve.
```
.. and waits for the flow to return

joeldudleyr3 replied:
Done

shamsasari commented on java/src/main/java/com/example/client/ExampleClientRPC.java:
```diff
@@ -3,52 +3,50 @@
-        Logger logger = LoggerFactory.getLogger(ExampleClientRPC.class);
-        HostAndPort nodeAddress = HostAndPort.fromString(args[0]);
-        CordaRPCClient client = new CordaRPCClient(nodeAddress, ConfigUtilities.configureTestSSL());
+        final Logger logger = LoggerFactory.getLogger(ExampleClientRPC.class);
```
Fields in Java are almost always private (the getter is public).

shamsasari commented on java/src/main/java/com/example/contract/PurchaseOrderState.java:
```diff
@@ -14,19 +15,22 @@
+    private final PurchaseOrderContract contract;
+    private final UniqueIdentifier linearId;
 
     public PurchaseOrderState(PurchaseOrder po,
```
`po` isn't descriptive

joeldudleyr3 replied:
Done

shamsasari commented on java/src/main/java/com/example/model/PurchaseOrder.java:
```diff
@@ -1,19 +1,24 @@
+ * A simple class representing a purchase order.
+ */
 public class PurchaseOrder {
     private int orderNumber;
```
Make these final

joeldudleyr3 replied:
I can't because the API uses the public empty constructor to reflect and set these. If I made these final, I'd have to create a constructor for them, which isn't necessary and would clutter the code.

shamsasari commented on java/src/main/java/com/example/model/PurchaseOrder.java:
```diff
@@ -30,10 +35,12 @@
+     * An inner class representing a purchase order address.
+     */
     public static class Address {
         private String city;
```
Likewise

joeldudleyr3 replied:
See above.

shamsasari commented on kotlin/src/main/kotlin/com/example/api/ExampleApi.kt:
```diff
@@ -2,19 +2,21 @@ package com.example.api
-class ExampleApi(val services: ServiceHub) {
-    val me: String = services.myInfo.legalIdentity.name
+class ExampleApi(val services: CordaRPCOps) {
+    val me: String = services.nodeIdentity().legalIdentity.name
```
myLegalName

joeldudleyr3 replied:
Done

shamsasari commented on kotlin/src/main/kotlin/com/example/api/ExampleApi.kt:
```diff
@@ -55,26 +57,36 @@ class ExampleApi(val services: ServiceHub) {
-        } else {
+    fun createPurchaseOrder(purchaseOrder: PurchaseOrder, @PathParam("party") partyName: String): Response {
+        val otherParty = services.partyFromName(partyName)
+        if(otherParty == null) {
```
Space after `if`

joeldudleyr3 replied:
Done
